test: mheap: add testcase

yingmingx opened the pull request:
This is an intergration testcase for mem_heap.
Add an testcase to verify that multiple threads
can share the same heap space without interfering
with each other.

Signed-off-by: Ying ming <mingx.ying@intel.com>

LeiW000 commented:
I think that we also need to malloc/free difference size of memory several times with multi-threads.

andyross commented on tests/kernel/mem_heap/mheap_api_concept/src/test_mheap_concept.c:
```diff
@@ -44,3 +51,49 @@ void test_mheap_malloc_align4(void)
+
+	for (int i = 0; i < THREAD_NUM; i++) {
+		/* verify free mheap in main thread */
+		k_free(block[i]);
```
k_free(NULL) is a documented noop, just like free(NULL) is in the C standard.  This is fine as long as block[i] always holds the return value from k_malloc().

LeiW000 replied:
@andyross, thanks. I was not aware of it.

andyross reviewed (approved):
Looks fine.  FWIW: if you really want to test for race conditions you're going to need to be doing thousands of allocations in a preemptable (and ideally SMP) context and not just one per thread.  But this isn't necessarily a stress test.

Commit:
test: mheap: add testcase

This is an intergration testcase for mem_heap.
Add an testcase to verify that multiple threads
can share the same heap space without interfering
with each other.

Signed-off-by: Ying ming <mingx.ying@intel.com>
